pdns/dnsdistdist/dnsdist-tcp-upstream.hh

```diff
@@ -0,0 +1,229 @@
+#pragma once
+
+#include "dolog.hh"
+
+class TCPClientThreadData
+{
+public:
+  TCPClientThreadData(): localRespRulactions(g_resprulactions.getLocal()), mplexer(std::unique_ptr<FDMultiplexer>(FDMultiplexer::getMultiplexerSilent()))
+  {
+  }
+
+  LocalHolders holders;
+  LocalStateHolder<vector<DNSDistResponseRuleAction> > localRespRulactions;
+  std::unique_ptr<FDMultiplexer> mplexer{nullptr};
+};
+
+struct ConnectionInfo
+{
+  ConnectionInfo(ClientState* cs_): cs(cs_), fd(-1)
+  {
+  }
+  ConnectionInfo(ConnectionInfo&& rhs): remote(rhs.remote), cs(rhs.cs), fd(rhs.fd)
+  {
+    rhs.cs = nullptr;
+    rhs.fd = -1;
+  }
+
+  ConnectionInfo(const ConnectionInfo& rhs) = delete;
+  ConnectionInfo& operator=(const ConnectionInfo& rhs) = delete;
+
+  ConnectionInfo& operator=(ConnectionInfo&& rhs)
+  {
+    remote = rhs.remote;
+    cs = rhs.cs;
+    rhs.cs = nullptr;
+    fd = rhs.fd;
+    rhs.fd = -1;
+    return *this;
+  }
+
+  ~ConnectionInfo()
+  {
+    if (fd != -1) {
+      close(fd);
+      fd = -1;
+    }
+    if (cs) {
+      --cs->tcpCurrentConnections;
```

Comment:
There's some asymmetry here; the increment happens in `tcpAcceptorThread`, but that's because it's only done after the accept I suppose?

Reply:
Yes, the object is allocated before the call to `accept()` so we can directly store the descriptor there, but we only want to increment after the connection has been accepted.